src/cascadia/TerminalApp/AppKeyBindings.cpp

```diff
@@ -1,158 +1,360 @@
+                        if (pair.second == commandString)
+                        {
+                            action = pair.first;
+                            parsedSuccessfully = true;
+                            break;
+                        }
+                    }
+                    if (!parsedSuccessfully)
+                    {
+                        continue;
+                    }
+                    // Try parsing the chord
+                    Settings::KeyChord chord{ false, false, false, 0 };
+                    parsedSuccessfully = false;
+                    try
+                    {
+                        chord = Settings::KeyChord::FromString(keyChordString);
+                        parsedSuccessfully = true;
+                    }
+                    catch (...)
+                    {
+                        continue;
+                    }
+                    if (parsedSuccessfully)
+                    {
+                        newBindings.SetKeyBinding(action, chord);
+                    }
+                }
+            }
+        }
+        return newBindings;
+    }
+
+    // Function Description:
+    // - Small helper to insert a given KeyChord, ShortcutAction pair into the
+    //   given json array
+    // Arguments:
+    // - bindingsArray: The JsonArray to insert the object into.
+    // - chord: The KeyChord to serailize and place in the json array
+    // - actionName: the name of the ShortcutAction to use with this KeyChord
+    void _AddShortcutToJsonArray(JsonArray bindingsArray,
+                                 const Settings::KeyChord& chord,
+                                 const std::wstring_view& actionName)
+    {
+        const auto keyString = chord.ToString();
+        if (keyString == L"")
+        {
+            return;
+        }
+
+        winrt::Windows::Data::Json::JsonObject jsonObject;
+        winrt::Windows::Data::Json::JsonArray keysArray;
+        keysArray.Append(JsonValue::CreateStringValue(keyString));
+        jsonObject.Insert(KEYS_KEY, keysArray);
+        jsonObject.Insert(COMMAND_KEY, JsonValue::CreateStringValue(actionName));
+
+        bindingsArray.Append(jsonObject);
+    }
+
+    // Method Description:
+    // - Serialize this AppKeyBindings to a json array of objects. Each object
+    //   in the array represents a single keybinding, mapping a KeyChord to a
+    //   ShortcutAction.
+    // Return Value:
+    // - a JsonArray which is an equivalent serialization of this object.
+    Windows::Data::Json::JsonArray AppKeyBindings::ToJson()
+    {
+        winrt::Windows::Data::Json::JsonArray bindingsArray;
+
+        // Iterate over all the possible actions in the names list, and see if
+        // it has a binding.
+        for (auto& actionName : commandNames)
+        {
+            const auto searchedForAction = actionName.first;
+            const auto searchedForName = actionName.second;
+            for (const auto& kv : _keyShortcuts)

```

Comment:
there just must be a better way to do this that isn't `O(N^2)`!

Reply:
derp that's literally the point of maps

Reply:
wait no, `_keyShortcuts` is `KeyChord->ShortcutAction`, while I have a name and a ShortcutAction. So I can't use the action to lookup the associated binding except by searching for it.

 I want to keep the ordering of the shortcuts in the json, so I can't iterate directly over all the `KeyChord, ShortcutAction` pairs in the `unordered_map` either